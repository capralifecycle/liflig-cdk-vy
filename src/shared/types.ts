/**
 * Shared type definitions for Vy custom resources
 */

/**
 * The different Vy environments
 */
export enum VyEnvironment {
  /**
   * Development environment
   */
  TEST = "test",

  /**
   * Production-like environment
   */
  STAGE = "stage",

  /**
   * Production environment
   */
  PROD = "prod",
}

export interface Scope {
  name: string
  description: string
}

export interface ResourceServer {
  identifier: string
  name: string
  scopes?: Scope[]
}

export interface ResourceServerUpdateRequest {
  identifier: string
  name: string
  scopes?: Scope[]
}

export interface AppClient {
  name: string
  scopes: string[]
  type: "frontend" | "backend"
  callback_urls: string[]
  logout_urls: string[]
  generate_secret?: boolean
  client_id?: string
  client_secret?: string
}

export interface AppClientUpdateRequest {
  name: string
  scopes: string[]
  callback_urls: string[]
  logout_urls: string[]
}

export interface DeploymentAccount {
  accountId: string
  slackChannel: string
}

export interface EnvironmentAccount {
  accountId: string
  ownerAccountId: string
}

export interface ArtifactVersion {
  uri: string
  store: string
  path: string
  version: string
}

export interface CognitoDetails {
  /**
   * The URL where users can authenticate
   */
  authUrl: string

  /**
   * The URL for the /.well-known/jwks.json endpoint
   */
  jwksUrl: string

  /**
   * The URL for the /.well-known/openid-configuration endpoint
   */
  openIdUrl: string

  /**
   * The URI for the issuer
   */
  issuer: string
}

/**
 * Custom Resource event types
 */
export interface CustomResourceRequest {
  RequestType: "Create" | "Update" | "Delete"
  RequestId: string
  ResponseURL: string
  ResourceType: string
  LogicalResourceId: string
  StackId: string
  PhysicalResourceId?: string
  ResourceProperties: Record<string, any>
  OldResourceProperties?: Record<string, any>
}

export interface CustomResourceResponse {
  Status: "SUCCESS" | "FAILED"
  Reason?: string
  PhysicalResourceId: string
  StackId: string
  RequestId: string
  LogicalResourceId: string
  Data?: Record<string, any>
}
