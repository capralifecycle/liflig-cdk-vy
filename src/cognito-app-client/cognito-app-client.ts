/**
 * CDK Construct for Cognito App Client
 */

import { createRequire } from "node:module"
import * as path from "node:path"
import { fileURLToPath } from "node:url"
import * as cdk from "aws-cdk-lib"
import { SecretValue } from "aws-cdk-lib"
import type * as iam from "aws-cdk-lib/aws-iam"
import type * as logs from "aws-cdk-lib/aws-logs"
import * as secretsmanager from "aws-cdk-lib/aws-secretsmanager"
import { Construct } from "constructs"
import type { AppClientProvider } from "../vy-cognito-provider"

const require = createRequire(import.meta.url)
const __filename = fileURLToPath(import.meta.url)
const __dirname = path.dirname(__filename)

/**
 * Type of app client
 */
export enum AppClientType {
  /**
   * Backend app client for machine-to-machine (M2M) authentication
   * Uses OAuth 2.0 Client Credentials grant type
   * Generates a client secret for authentication
   */
  BACKEND = "backend",

  /**
   * Frontend app client for user authentication
   * Uses OAuth 2.0 Authorization Code or Implicit grant types
   * Supports callback URLs and logout URLs for browser-based flows
   */
  FRONTEND = "frontend",
}

export interface CognitoAppClientProps {
  /**
   * An AppClientProvider provided from a VyCognitoProvider
   */
  readonly appClientProvider: AppClientProvider

  /**
   * The name of the app client
   * Must be unique within the environment
   */
  readonly name: string

  /**
   * The type of app client
   * - backend: Machine-to-machine authentication with client credentials
   * - frontend: User authentication with authorization code or implicit grant
   */
  readonly type: AppClientType

  /**
   * OAuth scopes for this client
   *
   * For backend clients: Use resource server scopes (e.g., 'https://api.vydev.io/read')
   * For frontend clients: Use OIDC scopes (e.g., 'email', 'openid', 'profile')
   *
   * @default - Empty array
   */
  readonly scopes?: string[]

  /**
   * Callback URLs for OAuth flows
   * Only used with frontend clients
   *
   * @default - No callback URLs
   */
  readonly callbackUrls?: string[]

  /**
   * Logout URLs for OAuth flows
   * Only used with frontend clients
   *
   * @default - No logout URLs
   */
  readonly logoutUrls?: string[]

  /**
   * Whether to generate a client secret
   * Automatically set based on type, but can be overridden
   *
   * @default - true for backend, false for frontend
   */
  readonly generateSecret?: boolean

  /**
   * Name of the secret in AWS Secrets Manager
   * Only applicable if generateSecret is true
   */
  readonly secretName?: string

  /**
   * Base domain for Cognito service
   * @default 'cognito.vydev.io'
   */
  readonly cognitoBaseDomain?: string

  /**
   * @default logs.RetentionDays.ONE_WEEK
   */
  readonly logsRetention?: logs.RetentionDays
}

/**
 * A Cognito App Client managed through Vy's central Cognito service
 *
 * App clients are the user pool authentication resources attached to your app.
 * Use an app client to configure the permitted authentication actions for an app.
 *
 * There are two types of app clients:
 * - **Backend**: Machine-to-machine authentication using OAuth 2.0 Client Credentials
 * - **Frontend**: User authentication using OAuth 2.0 Authorization Code or Implicit Grant
 *
 * @example
 * ```typescript
 *  // Create a VyCognitoProvider
 * const vyCognitoProvider = new VyCognitoProvider(this, 'MyProvider', {
 *   environment: VyEnvironment.TEST,
 * });
 *
 * // Backend app client for M2M authentication
 * const backendClient = new CognitoAppClient(this, 'BackendClient', {
 *   appClientProvider: vyCognitoProvider.appClientProvider,
 *   name: 'my-backend-service',
 *   type: AppClientType.BACKEND,
 *   scopes: ['https://api.vydev.io/read', 'https://api.vydev.io/write']
 * });
 *
 * // Access credentials
 * const clientId = backendClient.clientId;
 * const clientSecret = backendClient.clientSecret; // Stored in Secrets Manager
 * ```
 *
 * @example
 * ```typescript
 * // Create a VyCognitoProvider
 * const vyCognitoProvider = new VyCognitoProvider(this, 'MyProvider', {
 *   environment: VyEnvironment.TEST,
 * });
 *
 * // Frontend app client for user authentication
 * const frontendClient = new CognitoAppClient(this, 'FrontendClient', {
 *   appClientProvider: vyCognitoProvider.appClientProvider,
 *   name: 'my-web-app',
 *   type: AppClientType.FRONTEND,
 *   scopes: ['email', 'openid', 'profile'],
 *   callbackUrls: ['https://my-app.vydev.io/auth/callback'],
 *   logoutUrls: ['https://my-app.vydev.io/logout']
 * });
 * ```
 */
export class CognitoAppClient extends Construct {
  /**
   * The name of the app client
   */
  public readonly name: string

  /**
   * The underlying custom resource
   */
  public readonly resource: cdk.CustomResource

  /**
   * The client ID for authentication
   */
  public readonly clientId: string

  /**
   * A reference to the client secret
   */
  public readonly clientSecretArn?: string

  /**
   * The Secrets Manager secret containing the client secret
   */
  public readonly clientSecretSecret?: secretsmanager.ISecret

  constructor(scope: Construct, id: string, props: CognitoAppClientProps) {
    super(scope, id)

    this.name = props.name

    if (
      props.type === AppClientType.FRONTEND &&
      props.callbackUrls &&
      props.callbackUrls.length === 0
    ) {
      cdk.Annotations.of(this).addWarning(
        "Frontend app clients typically require callbackUrls for OAuth flows",
      )
    }

    // props.generateSecret overrides type-specific defaults
    // Default for backend clients is true; for frontend clients false
    const generateSecret =
      props.generateSecret ?? props.type === AppClientType.BACKEND

    this.resource = new cdk.CustomResource(this, "Resource", {
      serviceToken: props.appClientProvider.serviceToken,
      properties: {
        Environment: props.appClientProvider.environment,
        Name: props.name,
        Type: props.type,
        Scopes: props.scopes || [],
        CallbackUrls: props.callbackUrls || [],
        LogoutUrls: props.logoutUrls || [],
        GenerateSecret: generateSecret,
      },
      // Force replacement if type changes
      resourceType: "Custom::VyCognitoAppClient",
    })

    const client_id = this.resource.getAttString("ClientId")
    this.clientId = client_id

    if (generateSecret) {
      const client_secret = this.resource.getAttString("ClientSecret")
      const auth_url = props.appClientProvider.auth_url

      // For convenience, store auth_url, client_id, and client_secret in a single JSON secret
      const secretString = JSON.stringify({
        auth_url,
        client_id,
        client_secret,
      })

      this.clientSecretSecret = new secretsmanager.Secret(
        this,
        "ClientSecret",
        {
          secretName:
            props.secretName ??
            `/vy/${props.appClientProvider.environment}/cognito/app-client/${props.name}`,
          description: `auth_url, client_id and client_secret for Vy Cognito app client ${props.name}`,
          secretStringValue: SecretValue.unsafePlainText(secretString),
        },
      )

      this.clientSecretArn = this.clientSecretSecret.secretFullArn
    }
  }

  /**
   * Grant read access to the client secret (if it exists in Secrets Manager)
   */
  public grantReadSecret(grantee: iam.IGrantable): iam.Grant {
    if (!this.clientSecretSecret) {
      throw new Error("Client secret is not stored in Secrets Manager")
    }
    return this.clientSecretSecret.grantRead(grantee)
  }
}
