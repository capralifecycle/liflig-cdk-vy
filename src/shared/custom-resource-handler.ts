/**
 * Base handler for Custom Resources
 * Handles the CloudFormation response protocol
 */

import type { CustomResourceResponse } from "./types"

export function createSuccessResponse(
  physicalResourceId: string,
  data?: Record<string, any>,
): Pick<CustomResourceResponse, "Status" | "PhysicalResourceId" | "Data"> {
  return {
    Status: "SUCCESS",
    PhysicalResourceId: physicalResourceId,
    Data: data,
  }
}

export function createFailureResponse(
  physicalResourceId: string,
  reason: string,
): Pick<CustomResourceResponse, "Status" | "PhysicalResourceId" | "Reason"> {
  return {
    Status: "FAILED",
    PhysicalResourceId: physicalResourceId,
    Reason: reason,
  }
}

/**
 * Base error handler for custom resource operations
 */
export function handleError(error: unknown): string {
  if (error instanceof Error) {
    return error.message
  }
  return String(error)
}
