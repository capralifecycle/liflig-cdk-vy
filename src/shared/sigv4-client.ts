/**
 * Shared utility for making AWS SigV4 signed HTTP requests
 */

import { Sha256 } from "@aws-crypto/sha256-js"
import { defaultProvider } from "@aws-sdk/credential-provider-node"
import { HttpRequest } from "@smithy/protocol-http"
import { SignatureV4 } from "@smithy/signature-v4"

export interface SignedRequestOptions {
  method: string
  hostname: string
  path: string
  body?: string
  headers?: Record<string, string>
  region?: string
}

export interface SignedResponse {
  statusCode: number
  body: string
  headers: Record<string, string>
}

export async function signedRequest(
  options: SignedRequestOptions,
): Promise<SignedResponse> {
  const { method, hostname, path, body, headers = {}, region } = options

  const request = new HttpRequest({
    method,
    protocol: "https:",
    hostname,
    path,
    headers: {
      "Content-Type": "application/json",
      Host: hostname,
      ...headers,
    },
    body,
  })

  const credentialsProvider = defaultProvider()
  const credentials = await credentialsProvider()

  const signer = new SignatureV4({
    credentials,
    region: region ?? "eu-west-1",
    service: "execute-api",
    sha256: Sha256,
  })

  const signedRequest = await signer.sign(request)

  const url = `https://${signedRequest.hostname}${signedRequest.path}`
  const response = await fetch(url, {
    method: signedRequest.method,
    headers: signedRequest.headers as Record<string, string>,
    body: signedRequest.body,
  })

  const responseBody = await response.text()

  return {
    statusCode: response.status,
    body: responseBody,
    headers: Object.fromEntries(response.headers.entries()),
  }
}

/**
 * Helper to create base URL from environment
 */
export function createUrlFromEnvironment(
  baseUrl: string,
  urlPrefix: string,
  environment: string,
): string {
  if (environment === "prod") {
    return `${urlPrefix}.${baseUrl}`
  }
  return `${urlPrefix}.${environment}.${baseUrl}`
}
