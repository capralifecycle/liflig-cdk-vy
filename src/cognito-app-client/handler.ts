/**
 * Lambda handler for CognitoAppClient custom resource
 */

import {
  createFailureResponse,
  createSuccessResponse,
  handleError,
} from "../shared/custom-resource-handler"
import { createUrlFromEnvironment, signedRequest } from "../shared/sigv4-client"
import type {
  AppClient,
  AppClientUpdateRequest,
  CustomResourceRequest,
  CustomResourceResponse,
} from "../shared/types"

const COGNITO_BASE_DOMAIN =
  process.env.COGNITO_BASE_DOMAIN || "cognito.vydev.io"

interface AppClientProperties {
  Environment: string
  Name: string
  Type: "frontend" | "backend"
  Scopes?: string[]
  CallbackUrls?: string[]
  LogoutUrls?: string[]
  GenerateSecret?: boolean
}

async function createAppClient(
  baseUrl: string,
  client: AppClient,
): Promise<AppClient> {
  const response = await signedRequest({
    method: "POST",
    hostname: baseUrl,
    path: "/app-clients",
    body: JSON.stringify(client),
  })

  if (response.statusCode !== 201) {
    throw new Error(
      `Could not create resource: ${response.statusCode} - ${response.body}`,
    )
  }

  return JSON.parse(response.body)
}

async function readAppClient(
  baseUrl: string,
  name: string,
): Promise<AppClient> {
  const encodedName = encodeURIComponent(name)
  const response = await signedRequest({
    method: "GET",
    hostname: baseUrl,
    path: `/app-clients/${encodedName}`,
  })

  if (response.statusCode !== 200) {
    throw new Error(
      `Could not read resource: ${response.statusCode} - ${response.body}`,
    )
  }

  return JSON.parse(response.body)
}

async function updateAppClient(
  baseUrl: string,
  update: AppClientUpdateRequest,
): Promise<void> {
  const encodedName = encodeURIComponent(update.name)
  const response = await signedRequest({
    method: "PUT",
    hostname: baseUrl,
    path: `/app-clients/${encodedName}`,
    body: JSON.stringify(update),
  })

  if (response.statusCode !== 200) {
    throw new Error(
      `Could not update resource: ${response.statusCode} - ${response.body}`,
    )
  }
}

async function deleteAppClient(baseUrl: string, name: string): Promise<void> {
  const encodedName = encodeURIComponent(name)
  const response = await signedRequest({
    method: "DELETE",
    hostname: baseUrl,
    path: `/app-clients/${encodedName}`,
  })

  if (response.statusCode !== 200) {
    const message = `Could not delete resource: ${response.statusCode} - ${response.body}`

    if (response.statusCode === 404) {
      // Allow soft fail to avoid ROLLBACK_FAILED status
      console.warn(message)
    } else {
      throw new Error(message)
    }
  }
}

export async function handler(
  event: CustomResourceRequest,
): Promise<CustomResourceResponse> {
  const props = event.ResourceProperties as AppClientProperties
  const baseUrl = createUrlFromEnvironment(
    COGNITO_BASE_DOMAIN,
    "delegated",
    props.Environment,
  )

  try {
    switch (event.RequestType) {
      case "Create": {
        // We receive a string value for GenerateSecret, but we need a boolean
        // See https://github.com/aws-cloudformation/cloudformation-coverage-roadmap/issues/1037
        const generate_secret: boolean =
          typeof props.GenerateSecret === "boolean"
            ? props.GenerateSecret
            : props.GenerateSecret === "true"

        const client: AppClient = {
          name: props.Name,
          type: props.Type,
          scopes: props.Scopes || [],
          callback_urls: props.CallbackUrls || [],
          logout_urls: props.LogoutUrls || [],
          generate_secret,
        }

        const created = await createAppClient(baseUrl, client)

        return createSuccessResponse(event.PhysicalResourceId ?? created.name, {
          Name: created.name,
          ClientId: created.client_id || "",
          ClientSecret: created.client_secret || "",
          Type: created.type,
        }) as CustomResourceResponse
      }

      case "Update": {
        // Check if Type changed (requires replacement)
        const oldProps = event.OldResourceProperties as AppClientProperties
        if (oldProps && oldProps.Type !== props.Type) {
          throw new Error(
            "Cannot change app client type. This requires resource replacement.",
          )
        }

        const update: AppClientUpdateRequest = {
          name: props.Name,
          scopes: props.Scopes || [],
          callback_urls: props.CallbackUrls || [],
          logout_urls: props.LogoutUrls || [],
        }

        await updateAppClient(baseUrl, update)
        const updated = await readAppClient(baseUrl, props.Name)

        return createSuccessResponse(event.PhysicalResourceId ?? updated.name, {
          Name: updated.name,
          ClientId: updated.client_id || "",
          ClientSecret: updated.client_secret || "",
          Type: updated.type,
        }) as CustomResourceResponse
      }

      case "Delete": {
        const name = event.PhysicalResourceId || props.Name
        await deleteAppClient(baseUrl, name)

        return createSuccessResponse(name, {}) as CustomResourceResponse
      }
    }
  } catch (error) {
    console.error("Error:", error)
    return createFailureResponse(
      event.PhysicalResourceId || props.Name || "unknown",
      handleError(error),
    ) as CustomResourceResponse
  }
}
