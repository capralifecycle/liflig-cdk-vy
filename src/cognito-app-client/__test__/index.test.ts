import { App, Stack } from "aws-cdk-lib"
import "jest-cdk-snapshot"
import { VyEnvironment } from "../../shared/types"
import type { AppClientProvider } from "../../vy-cognito-provider"
import { AppClientType, CognitoAppClient } from "../cognito-app-client"

test("cognito frontend app client", () => {
  const app = new App()
  const stack = new Stack(app, "Stack")

  const mockAppClientProvider: AppClientProvider = {
    environment: VyEnvironment.TEST,
    serviceToken: "serviceToken",
    auth_url: "auth_url",
  }

  new CognitoAppClient(stack, "CognitoAppClient", {
    appClientProvider: mockAppClientProvider,
    name: "my-web-app",
    type: AppClientType.FRONTEND,
    scopes: ["email", "openid", "profile"],
    callbackUrls: ["https://my-app.vydev.io/auth/callback"],
    logoutUrls: ["https://my-app.vydev.io/logout"],
  })

  expect(stack).toMatchCdkSnapshot({
    ignoreAssets: true,
  })
})

test("cognito backend app client", () => {
  const app = new App()
  const stack = new Stack(app, "Stack")

  const mockAppClientProvider: AppClientProvider = {
    environment: VyEnvironment.STAGE,
    serviceToken: "serviceToken",
    auth_url: "auth_url",
  }

  new CognitoAppClient(stack, "CognitoAppClient", {
    appClientProvider: mockAppClientProvider,
    name: "my-backend-service",
    type: AppClientType.BACKEND,
    scopes: ["https://api.vydev.io/read", "https://api.vydev.io/write"],
  })

  expect(stack).toMatchCdkSnapshot({
    ignoreAssets: true,
  })
})

test("cognito frontend app client with custom secretName", () => {
  const app = new App()
  const stack = new Stack(app, "Stack")

  const mockAppClientProvider: AppClientProvider = {
    environment: VyEnvironment.PROD,
    serviceToken: "serviceToken",
    auth_url: "auth_url",
  }

  new CognitoAppClient(stack, "CognitoAppClient", {
    appClientProvider: mockAppClientProvider,
    name: "my-backend-service",
    type: AppClientType.BACKEND,
    scopes: ["https://api.vydev.io/read", "https://api.vydev.io/write"],
    secretName: "my-secret",
  })

  expect(stack).toMatchCdkSnapshot({
    ignoreAssets: true,
  })
})
