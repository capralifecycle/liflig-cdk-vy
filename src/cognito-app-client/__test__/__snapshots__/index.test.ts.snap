// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`cognito backend app client 1`] = `
Object {
  "Resources": Object {
    "CognitoAppClient6799695E": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "CallbackUrls": Array [],
        "Environment": "stage",
        "GenerateSecret": true,
        "LogoutUrls": Array [],
        "Name": "my-backend-service",
        "Scopes": Array [
          "https://api.vydev.io/read",
          "https://api.vydev.io/write",
        ],
        "ServiceToken": "serviceToken",
        "Type": "backend",
      },
      "Type": "Custom::VyCognitoAppClient",
      "UpdateReplacePolicy": "Delete",
    },
    "CognitoAppClientClientSecret0A69F29B": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "Description": "auth_url, client_id and client_secret for Vy Cognito app client my-backend-service",
        "Name": "/vy/stage/cognito/app-client/my-backend-service",
        "SecretString": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"auth_url\\":\\"auth_url\\",\\"client_id\\":\\"",
              Object {
                "Fn::GetAtt": Array [
                  "CognitoAppClient6799695E",
                  "ClientId",
                ],
              },
              "\\",\\"client_secret\\":\\"",
              Object {
                "Fn::GetAtt": Array [
                  "CognitoAppClient6799695E",
                  "ClientSecret",
                ],
              },
              "\\"}",
            ],
          ],
        },
      },
      "Type": "AWS::SecretsManager::Secret",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;

exports[`cognito frontend app client 1`] = `
Object {
  "Resources": Object {
    "CognitoAppClient6799695E": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "CallbackUrls": Array [
          "https://my-app.vydev.io/auth/callback",
        ],
        "Environment": "test",
        "GenerateSecret": false,
        "LogoutUrls": Array [
          "https://my-app.vydev.io/logout",
        ],
        "Name": "my-web-app",
        "Scopes": Array [
          "email",
          "openid",
          "profile",
        ],
        "ServiceToken": "serviceToken",
        "Type": "frontend",
      },
      "Type": "Custom::VyCognitoAppClient",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;

exports[`cognito frontend app client with custom secretName 1`] = `
Object {
  "Resources": Object {
    "CognitoAppClient6799695E": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "CallbackUrls": Array [],
        "Environment": "prod",
        "GenerateSecret": true,
        "LogoutUrls": Array [],
        "Name": "my-backend-service",
        "Scopes": Array [
          "https://api.vydev.io/read",
          "https://api.vydev.io/write",
        ],
        "ServiceToken": "serviceToken",
        "Type": "backend",
      },
      "Type": "Custom::VyCognitoAppClient",
      "UpdateReplacePolicy": "Delete",
    },
    "CognitoAppClientClientSecret0A69F29B": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "Description": "auth_url, client_id and client_secret for Vy Cognito app client my-backend-service",
        "Name": "my-secret",
        "SecretString": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"auth_url\\":\\"auth_url\\",\\"client_id\\":\\"",
              Object {
                "Fn::GetAtt": Array [
                  "CognitoAppClient6799695E",
                  "ClientId",
                ],
              },
              "\\",\\"client_secret\\":\\"",
              Object {
                "Fn::GetAtt": Array [
                  "CognitoAppClient6799695E",
                  "ClientSecret",
                ],
              },
              "\\"}",
            ],
          ],
        },
      },
      "Type": "AWS::SecretsManager::Secret",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;
